src/System.Management.Automation/engine/MshObject.cs

```diff
@@ -1144,8 +1144,7 @@ internal static string ToStringEnumerable(ExecutionContext context, IEnumerable
             {
                 if (obj != null)
                 {
-                    PSObject mshObj = PSObject.AsPSObject(obj);
-                    returnValue.Append(PSObject.ToString(context, mshObj, separator, format, formatProvider, false, false));
+                    returnValue.Append(PSObject.ToString(context, obj, separator, format, formatProvider, false, false));
```

Comment:
By wrapping the object into a `PSObject`, `PSObject.ToString` will be able to find the `ToString(...)` script method/code method defined for the type of the object in `types.ps1xml` files, so as to use that in precedence.
So simply removing this step will be a breaking change.

I looked at this issue myself, and currently `ToStringEnumerator` and `ToStringEnumerable` behave differently in terms of how the `ToString(...)` methods defined in `types.ps1xml` files are honored for the element object.
From the completeness perspective, `ToStringEnumerator` should do the same as `ToStringEnumerable` -- wrapping the object before passing to `PSObject.ToString`.

A simpler and narrow-scoped fix would be to special case string objects:
```c#
if (obj != null)
{
    object mshObj = obj -is string ? obj : PSObject.AsPSObject(obj);
    returnValue.Append(PSObject.ToString(context, mshObj, separator, format, formatProvider, false, false));
}
```

Reply:
My thoughts were that:
- having TryFastTrackPrimitiveTypes() implies that we don't want to find ToString() in types.ps1xml for _primitive_ types.
- it is unlikely breaking change. Why would anybody want to define custom ToString() for primitive types? I see no use. Even if so then I believe right way is to define a custom type with custom ToString().
- only `-Join` is affected and affected only on half as you mentioned - ToStringEnumerator() does not cast to PSObject. So we should fix either ToStringEnumerable() or ToStringEnumerator().

I tend to accept the unlikely breaking change and fix ToStringEnumerable(), not  ToStringEnumerator() - we get more benefits (great perf win as minimum).

Reply:
> Why would anybody want to define custom ToString() for primitive types? 

The thing is, you don't know the element of `enumerable` here is a primitive type. I can perfectly do `(Get-ChildItem c:\) -join ";"`.
For primitive types, I agree it's unlikely that a `ToString` method is defined in ETS for them, but they are more likely than `string` (yes, `string` is not a primitive type).

> only -Join is affected and affected only on half as you mentioned

This is not true. `PSObject.ToStringEnumerable` is used in `PSObject.ToString` recursively, and thus any calls to `PSObject.ToString` will be affected.

> ToStringEnumerator() does not cast to PSObject. So we should fix either ToStringEnumerable() or ToStringEnumerator().

`ToStringEnumerator()` is obviously the one that should be fixed. It's missing a functionality that powershell should have provided -- honor the custom `ToString` method defined in ETS for a type.

Reply:
Do you mean that the PR broke `(Get-ChildItem c:\) -join ";"` ? But this works as expected. Could you please show an example which the PR broke?

Reply:
Thanks for explanations!
If we call PSObject.ToString() for non-PSObject object we always falls in IFormattable code path as last resort and don't honor ETS - is it right?

The IFormattable code path doesn't allow me to fix PSObject.ToString() in universal way so I think we could call TryFastTrackPrimitiveTypes() in ToStringEnumerable() (and ToStringEnumerator()) before casting to PSObject and  calling PSObject.ToString().